sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -189,13 +205,44 @@ final class DataFrameWriter private[sql](df: DataFrame) {
         ifNotExists = false)).toRdd
   }
 
-  private def normalizedParCols: Option[Seq[String]] = partitioningColumns.map { parCols =>
-    parCols.map { col =>
-      df.logicalPlan.output
-        .map(_.name)
-        .find(df.sqlContext.analyzer.resolver(_, col))
-        .getOrElse(throw new AnalysisException(s"Partition column $col not found in existing " +
-          s"columns (${df.logicalPlan.output.map(_.name).mkString(", ")})"))
+  private def normalizedParCols: Option[Seq[String]] = partitioningColumns.map { cols =>
+    cols.map(normalize(_, "Partition"))
+  }
+
+  private def normalizedBucketCols: Option[Seq[String]] = bucketingColumns.map { cols =>
+    cols.map(normalize(_, "Bucketing"))
+  }
+
+  private def normalizedSortCols: Option[Seq[String]] = sortingColumns.map { cols =>
+    cols.map(normalize(_, "Sorting"))
+  }
+
+  private def getBucketSpec: Option[BucketSpec] = {
+    if (numBuckets.isEmpty && sortingColumns.isDefined) {
+      throw new IllegalArgumentException("Specify numBuckets and bucketing columns first.")
```

Comment:
This error message is not quite right. The order of bucketBy and sortBy doesn't matter right?

e.g.
df.write.bucketBy().sortBy() and df.write.sortBy.bucketBy both work?

Should be something like sortBy requires bucketBy.